<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>initialize ukbb — ukbb_init • BiocHail</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="initialize ukbb — ukbb_init"><meta property="og:description" content="initialize ukbb"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocHail</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.14</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/gwas_tut.html">01 BiocHail -- GWAS tutorial from hail.is</a>
    </li>
    <li>
      <a href="../articles/large_t2t.html">02 Working with larger VCF: T2T by chromosome</a>
    </li>
    <li>
      <a href="../articles/ukbb.html">03 Working with UK Biobank summary statistics</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/vjcitn/BiocHail/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>initialize ukbb</h1>
    <small class="dont-index">Source: <a href="https://github.com/vjcitn/BiocHail/blob/HEAD/R/ukbb_init.R" class="external-link"><code>R/ukbb_init.R</code></a></small>
    <div class="hidden name"><code>ukbb_init.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>initialize ukbb</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ukbb_init</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="value">
    <h2>Value</h2>
    

<p>python module reference</p>
    </div>
    <div id="note">
    <h2>Note</h2>
    <p>ukbb module may be passed around</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">ukbb</span> <span class="op">&lt;-</span> <span class="fu">ukbb_init</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ukbb</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [1] "AC_CUTOFFS"                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [2] "AF_CUTOFFS"                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [3] "CHROMOSOMES"                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [4] "ENGLAND_RECRUITMENT_CENTERS"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [5] "MIN_CASES"                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [6] "MIN_CASES_ALL"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [7] "MIN_CASES_EUR"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [8] "MISSENSE_CSQS"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   [9] "NULL_STR"                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [10] "NULL_STR_KEY"                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [11] "OTHER_CSQS"                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [12] "PHENO_COLUMN_FIELDS"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [13] "PHENO_DESCRIPTION_FIELDS"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [14] "PHENO_GWAS_FIELDS"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [15] "PHENO_KEY_FIELDS"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [16] "PILOT_PHENOTYPES"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [17] "PLOF_CSQS"                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [18] "POPS"                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [19] "P_THRESHOLDS"                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [20] "REFERENCE_GENOME"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [21] "SIG_THRESHOLD"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [22] "SYNONYMOUS_CSQS"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [23] "UKB_GNOMAD_POP_MAPPING"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [24] "add_coding_information"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [25] "add_white_noise_pheno"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [26] "add_whr"                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [27] "all_axis_join"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [28] "annotate_nearest_gene"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [29] "annotation_case_builder"                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [30] "annotation_case_builder_ukb_legacy"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [31] "annotations"                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [32] "brain_mri_data_path"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [33] "bucket"                                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [34] "check_timing_type"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [35] "check_trait_types"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [36] "coding_ht_path"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [37] "combine_datasets"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [38] "combine_pheno_files"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [39] "combine_pheno_files_multi_sex"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [40] "combine_pheno_files_multi_sex_legacy"         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [41] "combine_phenotypes"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [42] "combine_phenotypes_in_mt"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [43] "combine_phenotypes_with_name"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [44] "compute_cases_binary"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [45] "compute_n_cases"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [46] "conditional_phenotypes"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [47] "create_broadcast_dict"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [48] "create_gene_map_ht"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [49] "create_genome_intervals_file"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [50] "csv"                                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [51] "downsample_table_by_x_y"                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [52] "explode_by_p_threshold"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [53] "explode_lambda_ht"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [54] "filter_and_annotate_ukb_data"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [55] "filter_ht_for_plink"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [56] "filter_lambda_gc"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [57] "first_exposure_and_activity_monitor_data_path"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [58] "format_entries"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [59] "format_pheno_dir"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [60] "generate_lambda_ht_by_freq"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [61] "generate_qc_lambda_aggregator"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [62] "generic"                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [63] "genotype_bm_path"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [64] "genotype_loading"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [65] "genotype_samples_ht_path"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [66] "genotypes"                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [67] "get_age_sex_tsv_path"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [68] "get_all_codings"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [69] "get_analysis_data_path"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [70] "get_biomarker_ht_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [71] "get_cases_and_controls_from_log"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [72] "get_clump_sumstats_bm_path"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [73] "get_clump_sumstats_col_ht_path"               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [74] "get_clumping_results_path"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [75] "get_coding_path"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [76] "get_codings"                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [77] "get_covariates"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [78] "get_covariates_ht_path"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [79] "get_covid_data_path"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [80] "get_custom_pheno_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [81] "get_death_data_path"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [82] "get_files_in_parent_directory"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [83] "get_filtered_mt"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [84] "get_filtered_mt_with_x"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [85] "get_final_lambdas_path"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [86] "get_final_sample_set"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [87] "get_full_icd_data_description"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [88] "get_gene_intervals_path"                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [89] "get_gp_data_tsv_path"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [90] "get_gwas_sig_path"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [91] "get_heritability_dict"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [92] "get_heritability_from_log"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [93] "get_heritability_txt_path"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [94] "get_hesin_data_path"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [95] "get_hesin_delivery_ht_path"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [96] "get_hesin_mt_path"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [97] "get_hesin_raw_data_path"                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [98] "get_hm3_snplist_path"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  [99] "get_hq_samples"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [100] "get_inverse_normalize_status"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [101] "get_ld_matrix_path"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [102] "get_ld_score_flat_file_path"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [103] "get_ld_score_ht_path"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [104] "get_ld_variant_index_path"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [105] "get_meta_analysis_results_path"               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [106] "get_missing_codings"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [107] "get_modified_key"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [108] "get_n_even_intervals"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [109] "get_null_model_timing"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [110] "get_pheno_dict"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [111] "get_pheno_manifest_path"                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [112] "get_pheno_output_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [113] "get_phenos_to_run"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [114] "get_phenotype_results_qc_path"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [115] "get_phenotype_summary_path"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [116] "get_phesant_all_phenos_tsv_path"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [117] "get_phesant_reassignments"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [118] "get_prs_assess_ht_path"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [119] "get_prs_bm_path"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [120] "get_prs_mt_path"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [121] "get_pruned_tsv_path"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [122] "get_results_prefix"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [123] "get_results_timing_ht_path"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [124] "get_results_timing_tsv_path"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [125] "get_saige_timing_grep"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [126] "get_saige_version_from_log"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [127] "get_sample_file"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [128] "get_top_p_from_mt"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [129] "get_top_p_path"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [130] "get_ukb_additional_phenos_description_path"   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [131] "get_ukb_additional_phenos_tsv_path"           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [132] "get_ukb_af_ht_path"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [133] "get_ukb_grm_mt_path"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [134] "get_ukb_grm_plink_path"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [135] "get_ukb_grm_pruned_ht_path"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [136] "get_ukb_imputed_data"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [137] "get_ukb_meta"                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [138] "get_ukb_meta_pop_tsv_path"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [139] "get_ukb_pheno_ht_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [140] "get_ukb_pheno_mt"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [141] "get_ukb_pheno_mt_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [142] "get_ukb_phesant_summary_tsv_path"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [143] "get_ukb_samples_file_path"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [144] "get_ukb_sumstats_biomarkers_mt_path"          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [145] "get_ukb_sumstats_mt_path"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [146] "get_ukb_vep_path"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [147] "get_variant_results_path"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [148] "get_variant_results_qc_path"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [149] "get_vep_formatted_data"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [150] "hl"                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [151] "icd9_codings_ht_path"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [152] "icd9_codings_tsv_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [153] "icd_codings_ht_path"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [154] "icd_codings_tsv_path"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [155] "irnt"                                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [156] "join_pheno_hts_to_mt"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [157] "ld"                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [158] "load_activity_monitor_data"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [159] "load_all_mfi_data"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [160] "load_brain_mri_data"                          </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [161] "load_covid_data"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [162] "load_custom_pheno"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [163] "load_dob_ht"                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [164] "load_final_sumstats_mt"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [165] "load_first_occurrence_data"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [166] "load_gene_data"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [167] "load_hesin_data"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [168] "load_icd_data"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [169] "load_prescription_data"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [170] "load_showcase"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [171] "load_variant_data"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [172] "locus_alleles_to_chr_pos_ref_alt"             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [173] "mac_category_case_builder"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [174] "make_pairwise_ht"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [175] "mwzj_hts_by_tree"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [176] "os"                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [177] "pairwise_cooccurrence_ht_path"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [178] "pairwise_correlation_ht_path"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [179] "pheno_description_path"                       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [180] "pheno_description_raw_path"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [181] "pheno_folder"                                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [182] "pheno_ht_to_mt"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [183] "phenotype_loading"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [184] "phenotype_munging"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [185] "phenotypes"                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [186] "phesant_biomarker_phenotypes_tsv_path"        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [187] "phesant_folder"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [188] "post_process_gene_map_ht"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [189] "pre_phesant_biomarkers_tsv_path"              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [190] "pre_phesant_tsv_path"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [191] "pre_process_data_dictionary"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [192] "prescription_mapping_path"                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [193] "prescription_tsv_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [194] "public_bucket"                                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [195] "public_bucket_free"                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [196] "pull_out_col_keys"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [197] "pull_out_fields_from_entries"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [198] "re"                                           </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [199] "recode_pkd_to_legacy"                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [200] "recode_pkd_to_new"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [201] "recode_single_pkd_to_legacy"                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [202] "recode_single_pkd_to_new"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [203] "resources"                                    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [204] "results"                                      </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [205] "results_loading"                              </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [206] "separate_results_mt_by_pop"                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [207] "stringify_pheno_key_dict"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [208] "subprocess"                                   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [209] "summarize_data"                               </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [210] "temp_bucket"                                  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [211] "temp_bucket_7day"                             </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [212] "tempfile"                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [213] "ukb_imputed_bgen_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [214] "ukb_imputed_info_ht_path"                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [215] "ukb_imputed_info_path"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [216] "unify_saige_burden_ht_schema"                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [217] "unify_saige_ht_schema"                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [218] "unify_saige_ht_variant_schema"                </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [219] "union_ht"                                     </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [220] "union_hts_by_tree"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [221] "union_mts_by_tree"                            </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [222] "utils"                                        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [223] "uuid"                                         </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

