<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>03 Working with UK Biobank summary statistics • BiocHail</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="03 Working with UK Biobank summary statistics">
<meta property="og:description" content="BiocHail">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocHail</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/01_gwas_tut.html">01 BiocHail -- GWAS tutorial from hail.is</a>
    </li>
    <li>
      <a href="../articles/02_large_t2t.html">02 Working with larger VCF: T2T by chromosome</a>
    </li>
    <li>
      <a href="../articles/03_ukbb.html">03 Working with UK Biobank summary statistics</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/BiocHail/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="03_ukbb_files/htmlwidgets-1.6.1/htmlwidgets.js"></script><link href="03_ukbb_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="03_ukbb_files/datatables-binding-0.27/datatables.js"></script><link href="03_ukbb_files/dt-core-1.12.1/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="03_ukbb_files/dt-core-1.12.1/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="03_ukbb_files/dt-core-1.12.1/js/jquery.dataTables.min.js"></script><link href="03_ukbb_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="03_ukbb_files/crosstalk-1.2.0/js/crosstalk.min.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>03 Working with UK Biobank summary statistics</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">February 18, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/BiocHail/blob/HEAD/vignettes/03_ukbb.Rmd"><code>vignettes/03_ukbb.Rmd</code></a></small>
      <div class="hidden name"><code>03_ukbb.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this document we illustrate some approaches to working with UK Biobank summary statistics. Be sure that that the python module <code>ukbb_pan_ancestry</code> has been installed where reticulate can find it. (We don’t use basilisk as of 12/24/2022 because of issues in the terra spark cluster.)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/BiocHail">BiocHail</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="initialization-and-description">Initialization and description<a class="anchor" aria-label="anchor" href="#initialization-and-description"></a>
</h3>
<div class="section level4">
<h4 id="standalone">Standalone<a class="anchor" aria-label="anchor" href="#standalone"></a>
</h4>
<p>We have produced a representation of summary statistics for a sample of 9888 loci. This 5GB resource can be retrieved and cached with the following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hl</span> <span class="op">=</span> <span class="fu"><a href="../reference/hail_init.html">hail_init</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ss</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ukbb_sumstat_10kloci_mt.html">get_ukbb_sumstat_10kloci_mt</a></span><span class="op">(</span><span class="va">hl</span><span class="op">)</span> <span class="co"># can take about a minute to unzip 5GB</span></span>
<span><span class="va">ss</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="op">)</span>   <span class="co"># but if a persistent MatrixTable is at the location given</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] 9888</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] 7271</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>             <span class="co"># by env var HAIL_UKBB_SUMSTAT_10K_PATH it goes quickly</span></span></code></pre></div>
<p>To get a description of available content, we need a python chunk:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb5-1" title="1">r.ss.describe()</a></code></pre></div>
<pre><code><span><span class="co">## ----------------------------------------</span></span>
<span><span class="co">## Global fields:</span></span>
<span><span class="co">##     None</span></span>
<span><span class="co">## ----------------------------------------</span></span>
<span><span class="co">## Column fields:</span></span>
<span><span class="co">##     'trait_type': str</span></span>
<span><span class="co">##     'phenocode': str</span></span>
<span><span class="co">##     'pheno_sex': str</span></span>
<span><span class="co">##     'coding': str</span></span>
<span><span class="co">##     'modifier': str</span></span>
<span><span class="co">##     'pheno_data': array&lt;struct {</span></span>
<span><span class="co">##         n_cases: int32, </span></span>
<span><span class="co">##         n_controls: int32, </span></span>
<span><span class="co">##         heritability: float64, </span></span>
<span><span class="co">##         saige_version: str, </span></span>
<span><span class="co">##         inv_normalized: bool, </span></span>
<span><span class="co">##         pop: str</span></span>
<span><span class="co">##     }&gt;</span></span>
<span><span class="co">##     'description': str</span></span>
<span><span class="co">##     'description_more': str</span></span>
<span><span class="co">##     'coding_description': str</span></span>
<span><span class="co">##     'category': str</span></span>
<span><span class="co">##     'n_cases_full_cohort_both_sexes': int64</span></span>
<span><span class="co">##     'n_cases_full_cohort_females': int64</span></span>
<span><span class="co">##     'n_cases_full_cohort_males': int64</span></span>
<span><span class="co">## ----------------------------------------</span></span>
<span><span class="co">## Row fields:</span></span>
<span><span class="co">##     'locus': locus&lt;GRCh37&gt;</span></span>
<span><span class="co">##     'alleles': array&lt;str&gt;</span></span>
<span><span class="co">##     'gene': str</span></span>
<span><span class="co">##     'annotation': str</span></span>
<span><span class="co">## ----------------------------------------</span></span>
<span><span class="co">## Entry fields:</span></span>
<span><span class="co">##     'summary_stats': array&lt;struct {</span></span>
<span><span class="co">##         AF_Allele2: float64, </span></span>
<span><span class="co">##         imputationInfo: float64, </span></span>
<span><span class="co">##         BETA: float64, </span></span>
<span><span class="co">##         SE: float64, </span></span>
<span><span class="co">##         `p.value.NA`: float64, </span></span>
<span><span class="co">##         `AF.Cases`: float64, </span></span>
<span><span class="co">##         `AF.Controls`: float64, </span></span>
<span><span class="co">##         Pvalue: float64, </span></span>
<span><span class="co">##         low_confidence: bool</span></span>
<span><span class="co">##     }&gt;</span></span>
<span><span class="co">## ----------------------------------------</span></span>
<span><span class="co">## Column key: ['trait_type', 'phenocode', 'pheno_sex', 'coding', 'modifier']</span></span>
<span><span class="co">## Row key: ['locus', 'alleles']</span></span>
<span><span class="co">## ----------------------------------------</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="terra">Terra<a class="anchor" aria-label="anchor" href="#terra"></a>
</h4>
<p>Here’s a basic description of the summary stats table, with code that works in terra.bio:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hl</span> <span class="op">=</span> <span class="fu"><a href="../reference/bare_hail.html">bare_hail</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">hl</span><span class="op">$</span><span class="fu">init</span><span class="op">(</span>idempotent<span class="op">=</span><span class="cn">TRUE</span>, spark_conf<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">'spark.hadoop.fs.gs.requester.pays.mode'</span><span class="op">=</span> <span class="st">'CUSTOM'</span>,</span>
<span>  <span class="st">'spark.hadoop.fs.gs.requester.pays.buckets'</span><span class="op">=</span> <span class="st">'ukb-diverse-pops-public'</span>,</span>
<span>  <span class="st">'spark.hadoop.fs.gs.requester.pays.project.id'</span><span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"GOOGLE_PROJECT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We need to use a python chunk to get the output, using gs:// storage references.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb8-1" title="1">r.hl.read_matrix_table(<span class="st">'gs://ukb-diverse-pops-public/sumstats_release/results_full.mt'</span>).describe()</a></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="exploring-the-subset">Exploring the subset<a class="anchor" aria-label="anchor" href="#exploring-the-subset"></a>
</h3>
<div class="section level4">
<h4 id="metadata-on-study-phenotypes">Metadata on study phenotypes<a class="anchor" aria-label="anchor" href="#metadata-on-study-phenotypes"></a>
</h4>
<p>We’ll collect the column metadata to start to understand details of annotation of phenotypes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sscol</span> <span class="op">=</span> <span class="va">ss</span><span class="op">$</span><span class="fu">cols</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">collect</span><span class="op">(</span><span class="op">)</span> <span class="co"># OK because we are just working over column metadata</span></span>
<span><span class="va">ss1</span> <span class="op">=</span> <span class="va">sscol</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ss1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "annotate"                       "category"                      </span></span>
<span><span class="co">##  [3] "coding"                         "coding_description"            </span></span>
<span><span class="co">##  [5] "description"                    "description_more"              </span></span>
<span><span class="co">##  [7] "drop"                           "get"                           </span></span>
<span><span class="co">##  [9] "items"                          "keys"                          </span></span>
<span><span class="co">## [11] "modifier"                       "n_cases_full_cohort_both_sexes"</span></span>
<span><span class="co">## [13] "n_cases_full_cohort_females"    "n_cases_full_cohort_males"     </span></span>
<span><span class="co">## [15] "pheno_data"                     "pheno_sex"                     </span></span>
<span><span class="co">## [17] "phenocode"                      "select"                        </span></span>
<span><span class="co">## [19] "trait_type"                     "values"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ss1</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"phenocode"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "30600"</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ss1</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"description"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Albumin"</span></span></code></pre>
<p>We’ve introduced a function that extracts selected fields for a given phenotype, that accommodates availability of results for specific populations.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/multipop_df.html">multipop_df</a></span><span class="op">(</span><span class="va">ss1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## there are 6 pheno_data components, selecting component 1</span></span></code></pre>
<pre><code><span><span class="co">## there are 6 pheno_data components, selecting component 2</span></span></code></pre>
<pre><code><span><span class="co">## there are 6 pheno_data components, selecting component 3</span></span></code></pre>
<pre><code><span><span class="co">## there are 6 pheno_data components, selecting component 4</span></span></code></pre>
<pre><code><span><span class="co">## there are 6 pheno_data components, selecting component 5</span></span></code></pre>
<pre><code><span><span class="co">## there are 6 pheno_data components, selecting component 6</span></span></code></pre>
<pre><code><span><span class="co">##   trait_type phenocode description modifier coding_description coding n_cases</span></span>
<span><span class="co">## 1 biomarkers     30600     Albumin     irnt                 NA           5759</span></span>
<span><span class="co">## 2 biomarkers     30600     Albumin     irnt                 NA            856</span></span>
<span><span class="co">## 3 biomarkers     30600     Albumin     irnt                 NA           7694</span></span>
<span><span class="co">## 4 biomarkers     30600     Albumin     irnt                 NA           2340</span></span>
<span><span class="co">## 5 biomarkers     30600     Albumin     irnt                 NA         367192</span></span>
<span><span class="co">## 6 biomarkers     30600     Albumin     irnt                 NA           1364</span></span>
<span><span class="co">##   n_controls heritability pop</span></span>
<span><span class="co">## 1         NA   0.25412190 AFR</span></span>
<span><span class="co">## 2         NA   0.11268439 AMR</span></span>
<span><span class="co">## 3         NA   0.24110706 CSA</span></span>
<span><span class="co">## 4         NA   0.06126386 EAS</span></span>
<span><span class="co">## 5         NA   0.06449071 EUR</span></span>
<span><span class="co">## 6         NA   0.20458513 MID</span></span></code></pre>
<p>This can be iterated over all the elements of <code>sscol</code> to produce a comprehensive searchable table. Here’s a small illustration:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT" class="external-link">DT</a></span><span class="op">)</span></span>
<span><span class="va">ndf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sscol</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, <span class="va">multipop_df</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">ndf</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-6b560e151e6f77d24f5b" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-6b560e151e6f77d24f5b">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60"],["biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers","biomarkers"],["30600","30600","30600","30600","30600","30600","30610","30610","30610","30610","30610","30610","30620","30620","30620","30620","30620","30620","30630","30630","30630","30630","30630","30630","30640","30640","30640","30640","30640","30640","30650","30650","30650","30650","30650","30650","30660","30660","30660","30660","30660","30660","30670","30670","30670","30670","30670","30670","30680","30680","30680","30680","30680","30680","30690","30690","30690","30690","30690","30690"],["Albumin","Albumin","Albumin","Albumin","Albumin","Albumin","Alkaline phosphatase","Alkaline phosphatase","Alkaline phosphatase","Alkaline phosphatase","Alkaline phosphatase","Alkaline phosphatase","Alanine aminotransferase","Alanine aminotransferase","Alanine aminotransferase","Alanine aminotransferase","Alanine aminotransferase","Alanine aminotransferase","Apolipoprotein A","Apolipoprotein A","Apolipoprotein A","Apolipoprotein A","Apolipoprotein A","Apolipoprotein A","Apolipoprotein B","Apolipoprotein B","Apolipoprotein B","Apolipoprotein B","Apolipoprotein B","Apolipoprotein B","Aspartate aminotransferase","Aspartate aminotransferase","Aspartate aminotransferase","Aspartate aminotransferase","Aspartate aminotransferase","Aspartate aminotransferase","Direct bilirubin","Direct bilirubin","Direct bilirubin","Direct bilirubin","Direct bilirubin","Direct bilirubin","Urea","Urea","Urea","Urea","Urea","Urea","Calcium","Calcium","Calcium","Calcium","Calcium","Calcium","Cholesterol","Cholesterol","Cholesterol","Cholesterol","Cholesterol","Cholesterol"],["irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt","irnt"],[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],["","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""],[5759,856,7694,2340,367192,1364,6216,938,8422,2572,400988,1497,6214,937,8407,2570,400822,1499,5732,846,7679,2325,364987,1360,6169,932,8352,2553,399003,1496,6180,933,8395,2551,399482,1493,5165,744,6961,2159,340934,1235,6213,938,8422,2569,400687,1497,5758,856,7697,2344,367050,1361,6212,938,8422,2572,400963,1500],[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],[0.254121897246473,0.112684390050026,0.241107059208369,0.0612638623924412,0.0644907063086796,0.204585132400553,0.284713209524876,0.202000300691171,0.391926760575894,0,0.13052632331184,0.736078864452169,0.393531098684461,0,0.219169550459373,0.158296504215096,0.0627959578290811,0.205199438371898,0.30677607239906,0.333578943762955,0.357871888918728,0.290581087197675,0.121637999929837,0.483398782480507,0.254690364122439,0.364813591539292,0.221236551278177,0.185611051139175,0.0785137605607582,0,0.31480159882459,0.221399884523551,0.19039542293026,0,0.0720225122026994,0.321036779263955,0.340609517259067,0,0.286124940885441,0.162690359052229,0.104236568050214,0,0.152823191705072,0,0.162912991729159,0.136407968862874,0.0577579261020316,0,0.104765830617494,0,0.137741881658036,0.0374774232317139,0.0656627738181498,0,0.260847355647463,0.117187994922893,0.229808924377329,0.115794461086618,0.0637604095521063,0],["AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID","AFR","AMR","CSA","EAS","EUR","MID"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>trait_type<\/th>\n      <th>phenocode<\/th>\n      <th>description<\/th>\n      <th>modifier<\/th>\n      <th>coding_description<\/th>\n      <th>coding<\/th>\n      <th>n_cases<\/th>\n      <th>n_controls<\/th>\n      <th>heritability<\/th>\n      <th>pop<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[7,9]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="metadata-on-loci">Metadata on loci<a class="anchor" aria-label="anchor" href="#metadata-on-loci"></a>
</h4>
<p>We’ll trim the material to be worked with by sampling both rows and columns. (2023.01.08: In future revisions we will be able to control the seed for random sampling.)</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sss</span> <span class="op">=</span> <span class="va">ss</span><span class="op">$</span><span class="fu">sample_rows</span><span class="op">(</span><span class="fl">.01</span><span class="op">)</span><span class="op">$</span><span class="fu">sample_cols</span><span class="op">(</span><span class="fl">.01</span><span class="op">)</span></span>
<span><span class="va">sss</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] 107</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] 78</span></span></code></pre>
<p>Row metadata are simple to collect:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rss</span> <span class="op">=</span> <span class="va">sss</span><span class="op">$</span><span class="fu">rows</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">collect</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">rss1</span> <span class="op">=</span> <span class="va">rss</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rss1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "alleles"    "annotate"   "annotation" "drop"       "gene"      </span></span>
<span><span class="co">##  [6] "get"        "items"      "keys"       "locus"      "select"    </span></span>
<span><span class="co">## [11] "values"</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rss1</span><span class="op">$</span><span class="va">locus</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "contig"           "parse"            "position"         "reference_genome"</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rss1</span><span class="op">$</span><span class="va">locus</span><span class="op">$</span><span class="va">contig</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "1"</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"contig"</span>, <span class="st">"position"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">rss1</span><span class="op">$</span><span class="va">locus</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     contig   position </span></span>
<span><span class="co">##        "1" "26753527"</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h4>
<p>The summary statistics themselves reside in entries of the MatrixTable. This can be expensive to collect and so filtering methods beyond random sampling must be mastered. But here is a basic view.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sse</span> <span class="op">=</span> <span class="va">sss</span><span class="op">$</span><span class="fu">entries</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">collect</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 8346</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sse</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "alleles"                        "annotate"                      </span></span>
<span><span class="co">##  [3] "annotation"                     "category"                      </span></span>
<span><span class="co">##  [5] "coding"                         "coding_description"            </span></span>
<span><span class="co">##  [7] "description"                    "description_more"              </span></span>
<span><span class="co">##  [9] "drop"                           "gene"                          </span></span>
<span><span class="co">## [11] "get"                            "items"                         </span></span>
<span><span class="co">## [13] "keys"                           "locus"                         </span></span>
<span><span class="co">## [15] "modifier"                       "n_cases_full_cohort_both_sexes"</span></span>
<span><span class="co">## [17] "n_cases_full_cohort_females"    "n_cases_full_cohort_males"     </span></span>
<span><span class="co">## [19] "pheno_data"                     "pheno_sex"                     </span></span>
<span><span class="co">## [21] "phenocode"                      "select"                        </span></span>
<span><span class="co">## [23] "summary_stats"                  "trait_type"                    </span></span>
<span><span class="co">## [25] "values"</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sse1</span> <span class="op">=</span> <span class="va">sse</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>The <code>summary_stats</code> component has the association p-values – log10 transformed?</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sse1</span><span class="op">$</span><span class="va">summary_stats</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sse1</span><span class="op">$</span><span class="va">summary_stats</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "AF_Allele2"     "AF.Cases"       "AF.Controls"    "annotate"      </span></span>
<span><span class="co">##  [5] "BETA"           "drop"           "get"            "imputationInfo"</span></span>
<span><span class="co">##  [9] "items"          "keys"           "low_confidence" "p.value.NA"    </span></span>
<span><span class="co">## [13] "Pvalue"         "SE"             "select"         "values"</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sse1</span><span class="op">$</span><span class="va">summary_stats</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Pvalue</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] -0.340249</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<div class="section level3">
<h3 id="infrastructure">Infrastructure<a class="anchor" aria-label="anchor" href="#infrastructure"></a>
</h3>
<ul>
<li>Define an interface to this subset of 10k loci that supports queries like
<ul>
<li>has disease x been studied in UK Biobank?</li>
<li>how many phenotypes have been studied in K populations, K=2, 3, …?</li>
<li>how consistent is the annotation – are numbers of controls and cases always recorded?</li>
</ul>
</li>
<li>Do the UK Biobank portals produce information to resolve these questions?
<ul>
<li>if so, what are the API calls to obtain answers?</li>
<li>if not, what is missing from the portals to allow answers to be obtained?</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="substantive">Substantive<a class="anchor" aria-label="anchor" href="#substantive"></a>
</h3>
<ul>
<li>List the genes that are near the loci collected in the 10k random sample</li>
<li>(Hard) What is the most significant finding (position, phenotype, population) in the 10k subset?</li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
