<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>02 Working with larger VCF: T2T by chromosome • BiocHail</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="02 Working with larger VCF: T2T by chromosome">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BiocHail</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/gwas_tut.html">01 BiocHail -- GWAS tutorial from hail.is</a></li>
    <li><a class="dropdown-item" href="../articles/large_t2t.html">02 Working with larger VCF: T2T by chromosome</a></li>
    <li><a class="dropdown-item" href="../articles/ukbb.html">03 Working with UK Biobank summary statistics</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/vjcitn/BiocHail/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>02 Working with larger VCF: T2T by chromosome</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">October 30, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/BiocHail/blob/HEAD/vignettes/large_t2t.Rmd"><code>vignettes/large_t2t.Rmd</code></a></small>
      <div class="d-none name"><code>large_t2t.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this document we illustrate some issues with large data volumes.</p>
<p>Here is how we acquire the genotypes for the thousand genomes samples based on the T2T reference. We obtained bgzipped vcf via</p>
<pre><code><span><span class="fu">AnVIL</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnVIL/man/gsutil.html" class="external-link">gsutil_cp</a></span><span class="op">(</span><span class="st">"gs://fc-47de7dae-e8e6-429c-b760-b4ba49136eee/1KGP/joint_genotyping/joint_vcfs/raw/chr22.genotyped.vcf.gz"</span>, <span class="st">"."</span><span class="op">)</span></span></code></pre>
<p>Then we used hail in python to deal with the conversion to MatrixTable. We could have done this in R, but we had to learn how to manipulate the ‘reference sequence’. We have a conjectural ‘reference sequence’ json document in the json folder of the BiocHail package, used here as <code>t2tAnVIL.json</code>.</p>
<pre><code>&gt;&gt;&gt; import hail as h
&gt;&gt;&gt; rg = h.get_reference('GRCh38')
Initializing Hail with default parameters...
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Running on Apache Spark version 3.1.3
SparkUI available at http://756809c79837:4040
Welcome to
     __  __     &lt;&gt;__
    / /_/ /__  __/ /
   / __  / _ `/ / /
  /_/ /_/\_,_/_/_/   version 0.2.105-acd89e80c345
LOGGING: writing to /home/rstudio/hail-20221213-1558-0.2.105-acd89e80c345.log
&gt;&gt;&gt; nn = rg.read("t2tAnVIL.json")
&gt;&gt;&gt; h.import_vcf("chr22.genotyped.vcf.gz", force_bgz=True, reference_genome=nn).write("t2t22.mt")</code></pre>
<p>This operation seems to take a long time even with 64 cores. We could not get exact timing owing to some connectivity problems.</p>
<p>Here we’ll work with the genotype data for T2T chr17. We assume that the MatrixTable is located at the path given by the environment variable <code>HAIL_T2T_CHR17</code>. This MatrixTable is available in the Open Storage Network at osn:/bir190004-bucket01/Bioc1KGt2t/t17.zip. This is a 45 GB file. It should be unzipped at the location pointed to by <code>HAIL_T2T_CHR17</code>. Instructions for using rclone to acquire the zip file are given in the appendix.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/BiocHail">BiocHail</a></span><span class="op">)</span></span></code></pre></div>
<p>(If the above command indicates that BiocHail is not available, see the Installation section near the end of this document.)</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hail_init.html">hail_init</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># NB the following two commands are now encapsulated in the rg_update function</span></span>
<span><span class="va">nn</span> <span class="op">&lt;-</span> <span class="va">hl</span><span class="op">$</span><span class="fu">get_reference</span><span class="op">(</span><span class="st">'GRCh38'</span><span class="op">)</span></span>
<span><span class="va">nn</span> <span class="op">&lt;-</span> <span class="va">nn</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"json/t2tAnVIL.json"</span>, package<span class="op">=</span><span class="st">"BiocHail"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># updates the hail reference genome</span></span>
<span><span class="va">bigloc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"HAIL_T2T_CHR17"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">bigloc</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">mt17</span> <span class="op">&lt;-</span> <span class="va">hl</span><span class="op">$</span><span class="fu">read_matrix_table</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"HAIL_T2T_CHR17"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">mt17</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="population-stratification-assessment-via-pca">Population stratification assessment via PCA<a class="anchor" aria-label="anchor" href="#population-stratification-assessment-via-pca"></a>
</h2>
<p>The following command would compute PCs based on all SNP.</p>
<pre><code><span><span class="va">pcastuff</span> <span class="op">=</span> <span class="va">hl</span><span class="op">$</span><span class="fu">hwe_normalized_pca</span><span class="op">(</span><span class="va">mt17</span><span class="op">$</span><span class="va">GT</span><span class="op">)</span></span></code></pre>
<p>This seems impractical. We have found that with a 64 core machine at terra.bio, PCA on samples of 1-5% of the 3.8 million loci on T2T chr17 takes 40 minutes. Hail’s code seems good at utilizing all the cores.</p>
<p>We saved the PC scores for PCA based on 38k randomly sampled loci, and 191k randomly sampled loci. Here’s a simple view of the latter.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pcs_191k</span><span class="op">)</span></span>
<span><span class="fu">graphics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html" class="external-link">pairs</a></span><span class="op">(</span><span class="va">pcs_191k</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, pch<span class="op">=</span><span class="st">"."</span><span class="op">)</span></span></code></pre></div>
<p><img src="large_t2t_files/figure-html/lk191-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Comment on the gain in information about geographic origin achieved by using a 5% sample of loci instead of a 1% sample.</p></li>
<li><p>Find the geographic origins of donors of the 1000 genomes genotypes and bind them to <code>mt17</code> using the methods given in vignette <code>01_gwas_tut</code>. Use these codes to color the points in the PCA plot.</p></li>
<li><p>Produce an artificial “phenotype” for these donors via <code>rnorm(3202,0,1)</code>, bind it to the genotype data, and perform a naive GWAS. What are the loci most strongly associated with the artificial phenotype?</p></li>
<li><p>Produce a new artificial phenotype which has some association with geographic origin of donor, but no association with genotype. Produce a new naive GWAS, and a third using some of the PCA scores as covariates. What are the effects of this covariate adjustment on reasoning about genetic association with the artificial phenotype?</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="appendix-using-rclone-with-docker-to-get-the-chr17-data">Appendix: using rclone with docker to get the chr17 data<a class="anchor" aria-label="anchor" href="#appendix-using-rclone-with-docker-to-get-the-chr17-data"></a>
</h2>
<p>It can be painful to install and configure rclone. We use a docker container. Let RC_DATADIR be an environment variable evaluating to an available folder.</p>
<p>Also, place the text file with contents</p>
<pre><code>[osn]
type = s3
provider = AWS
endpoint = https://mghp.osn.xsede.org
acl = public
no_check_bucket = true</code></pre>
<p>in a file <code>rclone.conf</code> in a folder pointed to by the environment variable <code>RC_CONFDIR</code>.</p>
<p>Then the following</p>
<pre><code>docker run -v $RC_DATADIR:/data -v $RC_CONFDIR:/config/rclone -t rclone/rclone:latest ls osn:/bir190004-bucket01/Bioc1KGt2t</code></pre>
<p>will list the files with 1KG samples genotyped against the T2T reference.</p>
<p>Use the rclone <code>copyto</code> command to obtain a local copy of the zip file <code>t17.zip</code> in the folder pointed to by <code>$RC_DATADIR</code>:</p>
<pre><code>docker run -v $RC_DATADIR:/data -v $RC_CONFDIR:/config/rclone -t rclone/rclone:latest copyto osn:/bir190004-bucket01/Bioc1KGt2t/t17.zip ./t17.zip</code></pre>
<p>This file should be unzipped in a folder to which the environment variable <code>HAIL_T2T_CHR17</code> will point.</p>
</div>
<div class="section level2">
<h2 id="installing-biochail">Installing <code>BiocHail</code><a class="anchor" aria-label="anchor" href="#installing-biochail"></a>
</h2>
<p><code>BiocHail</code> should be installed as follows:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"BiocHail"</span><span class="op">)</span></span></code></pre></div>
<p>As of 1.0.0, a JDK for Java version <code>&lt;=</code> 11.0 is necessary to use the version of Hail that is installed with the package. This package should be usable on MacOS with suitable java support. If Java version <code>&gt;=</code> 8.x is used, warnings from Apache Spark may be observed. To the best of our knowledge the conditions to which the warnings pertain do not affect program performance.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.7</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] BiocHail_1.7.0   BiocStyle_2.32.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] sass_0.4.9            utf8_1.2.4            generics_0.1.3       </span></span>
<span><span class="co">##  [4] lattice_0.22-6        RSQLite_2.3.7         digest_0.6.37        </span></span>
<span><span class="co">##  [7] magrittr_2.0.3        grid_4.4.1            evaluate_1.0.1       </span></span>
<span><span class="co">## [10] bookdown_0.40         fastmap_1.2.0         blob_1.2.4           </span></span>
<span><span class="co">## [13] Matrix_1.7-0          jsonlite_1.8.9        DBI_1.2.3            </span></span>
<span><span class="co">## [16] BiocManager_1.30.25   httr_1.4.7            fansi_1.0.6          </span></span>
<span><span class="co">## [19] textshaping_0.4.0     jquerylib_0.1.4       cli_3.6.3            </span></span>
<span><span class="co">## [22] rlang_1.1.4           dbplyr_2.5.0          basilisk.utils_1.16.0</span></span>
<span><span class="co">## [25] bit64_4.5.2           withr_3.0.1           cachem_1.1.0         </span></span>
<span><span class="co">## [28] yaml_2.3.10           parallel_4.4.1        tools_4.4.1          </span></span>
<span><span class="co">## [31] dir.expiry_1.12.0     memoise_2.0.1         dplyr_1.1.4          </span></span>
<span><span class="co">## [34] filelock_1.0.3        basilisk_1.16.0       BiocGenerics_0.50.0  </span></span>
<span><span class="co">## [37] reticulate_1.39.0     curl_5.2.3            png_0.1-8            </span></span>
<span><span class="co">## [40] vctrs_0.6.5           R6_2.5.1              BiocFileCache_2.12.0 </span></span>
<span><span class="co">## [43] lifecycle_1.0.4       fs_1.6.4              htmlwidgets_1.6.4    </span></span>
<span><span class="co">## [46] bit_4.5.0             ragg_1.3.3            pkgconfig_2.0.3      </span></span>
<span><span class="co">## [49] desc_1.4.3            pkgdown_2.1.1         pillar_1.9.0         </span></span>
<span><span class="co">## [52] bslib_0.8.0           Rcpp_1.0.13           glue_1.8.0           </span></span>
<span><span class="co">## [55] systemfonts_1.1.0     highr_0.11            xfun_0.48            </span></span>
<span><span class="co">## [58] tibble_3.2.1          tidyselect_1.2.1      knitr_1.48           </span></span>
<span><span class="co">## [61] htmltools_0.5.8.1     rmarkdown_2.28        compiler_4.4.1</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
