<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>03 Working with UK Biobank summary statistics • BiocHail</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="03 Working with UK Biobank summary statistics">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">BiocHail</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/gwas_tut.html">01 BiocHail -- GWAS tutorial from hail.is</a></li>
    <li><a class="dropdown-item" href="../articles/large_t2t.html">02 Working with larger VCF: T2T by chromosome</a></li>
    <li><a class="dropdown-item" href="../articles/ukbb.html">03 Working with UK Biobank summary statistics</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="nav-link" href="https://github.com/vjcitn/BiocHail/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>03 Working with UK Biobank summary statistics</h1>
                        <h4 data-toc-skip class="author">Vincent J. Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">October 30, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/BiocHail/blob/HEAD/vignettes/ukbb.Rmd"><code>vignettes/ukbb.Rmd</code></a></small>
      <div class="d-none name"><code>ukbb.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this document we illustrate some approaches to working with UK Biobank summary statistics. Be sure that that the python module <code>ukbb_pan_ancestry</code> has been installed where reticulate can find it. (We don’t use basilisk as of 12/24/2022 because of issues in the terra spark cluster.)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/BiocHail">BiocHail</a></span><span class="op">)</span></span></code></pre></div>
<p>If the above command indicates that BiocHail is not available, see the Installation section near the end of this document.</p>
<div class="section level3">
<h3 id="initialization-and-description">Initialization and description<a class="anchor" aria-label="anchor" href="#initialization-and-description"></a>
</h3>
<div class="section level4">
<h4 id="standalone">Standalone<a class="anchor" aria-label="anchor" href="#standalone"></a>
</h4>
<p>We have produced a representation of summary statistics for a sample of 9888 loci. This 5GB resource can be retrieved and cached with the following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hl</span> <span class="op">=</span> <span class="fu"><a href="../reference/hail_init.html">hail_init</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ss</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_ukbb_sumstat_10kloci_mt.html">get_ukbb_sumstat_10kloci_mt</a></span><span class="op">(</span><span class="va">hl</span><span class="op">)</span> <span class="co"># can take about a minute to unzip 5GB</span></span>
<span><span class="va">ss</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="op">)</span>   <span class="co"># but if a persistent MatrixTable is at the location given</span></span>
<span>             <span class="co"># by env var HAIL_UKBB_SUMSTAT_10K_PATH it goes quickly</span></span></code></pre></div>
<p>To get a description of available content, we need a python chunk:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>r.ss.describe()</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="terra">Terra<a class="anchor" aria-label="anchor" href="#terra"></a>
</h4>
<p>Here’s a basic description of the summary stats table, with code that works in terra.bio:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hl</span> <span class="op">=</span> <span class="fu"><a href="../reference/bare_hail.html">bare_hail</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">hl</span><span class="op">$</span><span class="fu">init</span><span class="op">(</span>idempotent<span class="op">=</span><span class="cn">TRUE</span>, spark_conf<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">'spark.hadoop.fs.gs.requester.pays.mode'</span><span class="op">=</span> <span class="st">'CUSTOM'</span>,</span>
<span>  <span class="st">'spark.hadoop.fs.gs.requester.pays.buckets'</span><span class="op">=</span> <span class="st">'ukb-diverse-pops-public'</span>,</span>
<span>  <span class="st">'spark.hadoop.fs.gs.requester.pays.project.id'</span><span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"GOOGLE_PROJECT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We need to use a python chunk to get the output, using gs:// storage references.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>r.hl.read_matrix_table(<span class="st">'gs://ukb-diverse-pops-public/sumstats_release/results_full.mt'</span>).describe()</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="exploring-the-subset">Exploring the subset<a class="anchor" aria-label="anchor" href="#exploring-the-subset"></a>
</h3>
<div class="section level4">
<h4 id="metadata-on-study-phenotypes">Metadata on study phenotypes<a class="anchor" aria-label="anchor" href="#metadata-on-study-phenotypes"></a>
</h4>
<p>We’ll collect the column metadata to start to understand details of annotation of phenotypes.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sscol</span> <span class="op">=</span> <span class="va">ss</span><span class="op">$</span><span class="fu">cols</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">collect</span><span class="op">(</span><span class="op">)</span> <span class="co"># OK because we are just working over column metadata</span></span>
<span><span class="va">ss1</span> <span class="op">=</span> <span class="va">sscol</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ss1</span><span class="op">)</span></span>
<span><span class="va">ss1</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"phenocode"</span><span class="op">)</span></span>
<span><span class="va">ss1</span><span class="op">$</span><span class="fu">get</span><span class="op">(</span><span class="st">"description"</span><span class="op">)</span></span></code></pre></div>
<p>We’ve introduced a function that extracts selected fields for a given phenotype, that accommodates availability of results for specific populations.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/multipop_df.html">multipop_df</a></span><span class="op">(</span><span class="va">ss1</span><span class="op">)</span></span></code></pre></div>
<p>This can be iterated over all the elements of <code>sscol</code> to produce a comprehensive searchable table. Here’s a small illustration:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT" class="external-link">DT</a></span><span class="op">)</span></span>
<span><span class="va">ndf</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sscol</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>, <span class="va">multipop_df</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">ndf</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="metadata-on-loci">Metadata on loci<a class="anchor" aria-label="anchor" href="#metadata-on-loci"></a>
</h4>
<p>We’ll trim the material to be worked with by sampling both rows and columns. (2023.01.08: In future revisions we will be able to control the seed for random sampling.)</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sss</span> <span class="op">=</span> <span class="va">ss</span><span class="op">$</span><span class="fu">sample_rows</span><span class="op">(</span><span class="fl">.01</span><span class="op">)</span><span class="op">$</span><span class="fu">sample_cols</span><span class="op">(</span><span class="fl">.01</span><span class="op">)</span></span>
<span><span class="va">sss</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Row metadata are simple to collect:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rss</span> <span class="op">=</span> <span class="va">sss</span><span class="op">$</span><span class="fu">rows</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">collect</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">rss1</span> <span class="op">=</span> <span class="va">rss</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rss1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rss1</span><span class="op">$</span><span class="va">locus</span><span class="op">)</span></span>
<span><span class="va">rss1</span><span class="op">$</span><span class="va">locus</span><span class="op">$</span><span class="va">contig</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"contig"</span>, <span class="st">"position"</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">rss1</span><span class="op">$</span><span class="va">locus</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h4>
<p>The summary statistics themselves reside in entries of the MatrixTable. This can be expensive to collect and so filtering methods beyond random sampling must be mastered. But here is a basic view.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sse</span> <span class="op">=</span> <span class="va">sss</span><span class="op">$</span><span class="fu">entries</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">collect</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sse</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sse</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sse1</span> <span class="op">=</span> <span class="va">sse</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>The <code>summary_stats</code> component has the association p-values – log10 transformed?</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sse1</span><span class="op">$</span><span class="va">summary_stats</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sse1</span><span class="op">$</span><span class="va">summary_stats</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">sse1</span><span class="op">$</span><span class="va">summary_stats</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">Pvalue</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<div class="section level3">
<h3 id="infrastructure">Infrastructure<a class="anchor" aria-label="anchor" href="#infrastructure"></a>
</h3>
<ul>
<li>Define an interface to this subset of 10k loci that supports queries like
<ul>
<li>has disease x been studied in UK Biobank?</li>
<li>how many phenotypes have been studied in K populations, K=2, 3, …?</li>
<li>how consistent is the annotation – are numbers of controls and cases always recorded?</li>
</ul>
</li>
<li>Do the UK Biobank portals produce information to resolve these questions?
<ul>
<li>if so, what are the API calls to obtain answers?</li>
<li>if not, what is missing from the portals to allow answers to be obtained?</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="substantive">Substantive<a class="anchor" aria-label="anchor" href="#substantive"></a>
</h3>
<ul>
<li>List the genes that are near the loci collected in the 10k random sample</li>
<li>(Hard) What is the most significant finding (position, phenotype, population) in the 10k subset?</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="installing-biochail">Installing <code>BiocHail</code><a class="anchor" aria-label="anchor" href="#installing-biochail"></a>
</h2>
<p><code>BiocHail</code> should be installed as follows:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"BiocHail"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="sessioninfo">SessionInfo<a class="anchor" aria-label="anchor" href="#sessioninfo"></a>
</h2>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.7</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] BiocStyle_2.32.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] cli_3.6.3           knitr_1.48          rlang_1.1.4        </span></span>
<span><span class="co">##  [4] xfun_0.48           png_0.1-8           textshaping_0.4.0  </span></span>
<span><span class="co">##  [7] jsonlite_1.8.9      htmltools_0.5.8.1   ragg_1.3.3         </span></span>
<span><span class="co">## [10] sass_0.4.9          rmarkdown_2.28      grid_4.4.1         </span></span>
<span><span class="co">## [13] evaluate_1.0.1      jquerylib_0.1.4     fastmap_1.2.0      </span></span>
<span><span class="co">## [16] yaml_2.3.10         lifecycle_1.0.4     bookdown_0.40      </span></span>
<span><span class="co">## [19] BiocManager_1.30.25 compiler_4.4.1      fs_1.6.4           </span></span>
<span><span class="co">## [22] Rcpp_1.0.13         htmlwidgets_1.6.4   lattice_0.22-6     </span></span>
<span><span class="co">## [25] systemfonts_1.1.0   digest_0.6.37       R6_2.5.1           </span></span>
<span><span class="co">## [28] reticulate_1.39.0   bslib_0.8.0         Matrix_1.7-0       </span></span>
<span><span class="co">## [31] tools_4.4.1         pkgdown_2.1.1       cachem_1.1.0       </span></span>
<span><span class="co">## [34] desc_1.4.3</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
