<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>pheno_data component harvesting from columns of summary stats MatrixTable
allowing for info on multiple populations in the pheno_data component — multipop_df • BiocHail</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="pheno_data component harvesting from columns of summary stats MatrixTable
allowing for info on multiple populations in the pheno_data component — multipop_df"><meta property="og:description" content="pheno_data component harvesting from columns of summary stats MatrixTable
allowing for info on multiple populations in the pheno_data component"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocHail</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/gwas_tut.html">01 BiocHail -- GWAS tutorial from hail.is</a>
    </li>
    <li>
      <a href="../articles/large_t2t.html">02 Working with larger VCF: T2T by chromosome</a>
    </li>
    <li>
      <a href="../articles/ukbb.html">03 Working with UK Biobank summary statistics</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/vjcitn/BiocHail/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>pheno_data component harvesting from columns of summary stats MatrixTable
allowing for info on multiple populations in the pheno_data component</h1>
    <small class="dont-index">Source: <a href="https://github.com/vjcitn/BiocHail/blob/HEAD/R/pheno_col_ops.R"><code>R/pheno_col_ops.R</code></a></small>
    <div class="hidden name"><code>multipop_df.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>pheno_data component harvesting from columns of summary stats MatrixTable
allowing for info on multiple populations in the pheno_data component</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">multipop_df</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  top2get <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"trait_type"</span>, <span class="st">"phenocode"</span>, <span class="st">"description"</span>, <span class="st">"modifier"</span>, <span class="st">"coding_description"</span>,</span>
<span>    <span class="st">"coding"</span><span class="op">)</span>,</span>
<span>  pheno2get <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n_cases"</span>, <span class="st">"n_controls"</span>, <span class="st">"heritability"</span>, <span class="st">"pop"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>x</dt>
<dd><p>Struct - a single element of the list returned by mt$cols()$collect()</p></dd>


<dt>top2get</dt>
<dd><p>character() vector of general fields to retrieve</p></dd>


<dt>pheno2get</dt>
<dd><p>character() vector of fields to be retrieved for each subpopulation</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>data.frame</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"HAIL_UKBB_SUMSTAT_10K_PATH"</span><span class="op">)</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="va">hl</span> <span class="op">=</span> <span class="fu"><a href="hail_init.html">hail_init</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">ss</span> <span class="op">=</span> <span class="fu"><a href="get_ukbb_sumstat_10kloci_mt.html">get_ukbb_sumstat_10kloci_mt</a></span><span class="op">(</span><span class="va">hl</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="va">sscol</span> <span class="op">=</span> <span class="va">ss</span><span class="op">$</span><span class="fu">cols</span><span class="op">(</span><span class="op">)</span><span class="op">$</span><span class="fu">collect</span><span class="op">(</span><span class="op">)</span> <span class="co"># may take a bit of time</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sscol</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu">multipop_df</span><span class="op">(</span><span class="va">sscol</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 7271</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> there are 6 pheno_data components, selecting component 1</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> there are 6 pheno_data components, selecting component 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> there are 6 pheno_data components, selecting component 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> there are 6 pheno_data components, selecting component 4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> there are 6 pheno_data components, selecting component 5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> there are 6 pheno_data components, selecting component 6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   trait_type phenocode description modifier coding_description coding n_cases</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 biomarkers     30600     Albumin     irnt                 NA           5759</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 biomarkers     30600     Albumin     irnt                 NA            856</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 biomarkers     30600     Albumin     irnt                 NA           7694</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 biomarkers     30600     Albumin     irnt                 NA           2340</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 biomarkers     30600     Albumin     irnt                 NA         367192</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 biomarkers     30600     Albumin     irnt                 NA           1364</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   n_controls heritability pop</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1         NA   0.25412190 AFR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2         NA   0.11268439 AMR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3         NA   0.24110706 CSA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         NA   0.06126386 EAS</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         NA   0.06449071 EUR</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6         NA   0.20458513 MID</span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># to get an overview of all phenotype-cohort combinations in a searchable table</span></span></span>
<span class="r-in"><span><span class="va">mmm</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sscol</span>, <span class="va">multipop_df</span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mymy</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">mmm</span><span class="op">)</span> <span class="co"># over 16k rows</span></span></span>
<span class="r-in"><span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">mymy</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

