<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>02 Working with larger VCF: T2T by chromosome • BiocHail</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="02 Working with larger VCF: T2T by chromosome">
<meta property="og:description" content="BiocHail">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">BiocHail</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/gwas_tut.html">01 BiocHail -- GWAS tutorial from hail.is</a>
    </li>
    <li>
      <a href="../articles/large_t2t.html">02 Working with larger VCF: T2T by chromosome</a>
    </li>
    <li>
      <a href="../articles/ukbb.html">03 Working with UK Biobank summary statistics</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/vjcitn/BiocHail/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>02 Working with larger VCF: T2T by
chromosome</h1>
                        <h4 data-toc-skip class="author">Vincent J.
Carey, stvjc at channing.harvard.edu</h4>
            
            <h4 data-toc-skip class="date">March 08, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/vjcitn/BiocHail/blob/HEAD/vignettes/large_t2t.Rmd" class="external-link"><code>vignettes/large_t2t.Rmd</code></a></small>
      <div class="hidden name"><code>large_t2t.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this document we illustrate some issues with large data
volumes.</p>
<p>Here is how we acquire the genotypes for the thousand genomes samples
based on the T2T reference. We obtained bgzipped vcf via</p>
<pre><code><span><span class="fu">AnVIL</span><span class="fu">::</span><span class="fu">gsutil_cp</span><span class="op">(</span><span class="st">"gs://fc-47de7dae-e8e6-429c-b760-b4ba49136eee/1KGP/joint_genotyping/joint_vcfs/raw/chr22.genotyped.vcf.gz"</span>, <span class="st">"."</span><span class="op">)</span></span></code></pre>
<p>Then we used hail in python to deal with the conversion to
MatrixTable. We could have done this in R, but we had to learn how to
manipulate the ‘reference sequence’. We have a conjectural ‘reference
sequence’ json document in the json folder of the BiocHail package, used
here as <code>t2tAnVIL.json</code>.</p>
<pre><code>&gt;&gt;&gt; import hail as h
&gt;&gt;&gt; rg = h.get_reference('GRCh38')
Initializing Hail with default parameters...
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Running on Apache Spark version 3.1.3
SparkUI available at http://756809c79837:4040
Welcome to
     __  __     &lt;&gt;__
    / /_/ /__  __/ /
   / __  / _ `/ / /
  /_/ /_/\_,_/_/_/   version 0.2.105-acd89e80c345
LOGGING: writing to /home/rstudio/hail-20221213-1558-0.2.105-acd89e80c345.log
&gt;&gt;&gt; nn = rg.read("t2tAnVIL.json")
&gt;&gt;&gt; h.import_vcf("chr22.genotyped.vcf.gz", force_bgz=True, reference_genome=nn).write("t2t22.mt")</code></pre>
<p>This operation seems to take a long time even with 64 cores. We could
not get exact timing owing to some connectivity problems.</p>
<p>Here we’ll work with the genotype data for T2T chr17. We assume that
the MatrixTable is located at the path given by the environment variable
<code>HAIL_T2T_CHR17</code>. This MatrixTable is available in the Open
Storage Network at osn:/bir190004-bucket01/Bioc1KGt2t/t17.zip. This is a
45 GB file. It should be unzipped at the location pointed to by
<code>HAIL_T2T_CHR17</code>. Instructions for using rclone to acquire
the zip file are given in the appendix.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vjcitn/BiocHail" class="external-link">BiocHail</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hl</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hail_init.html">hail_init</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># NB the following two commands are now encapsulated in the rg_update function</span></span>
<span><span class="va">nn</span> <span class="op">&lt;-</span> <span class="va">hl</span><span class="op">$</span><span class="fu">get_reference</span><span class="op">(</span><span class="st">'GRCh38'</span><span class="op">)</span></span>
<span><span class="va">nn</span> <span class="op">&lt;-</span> <span class="va">nn</span><span class="op">$</span><span class="fu">read</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"json/t2tAnVIL.json"</span>, package<span class="op">=</span><span class="st">"BiocHail"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># updates the hail reference genome</span></span>
<span><span class="va">bigloc</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"HAIL_T2T_CHR17"</span><span class="op">)</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">bigloc</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">mt17</span> <span class="op">&lt;-</span> <span class="va">hl</span><span class="op">$</span><span class="fu">read_matrix_table</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"HAIL_T2T_CHR17"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">mt17</span><span class="op">$</span><span class="fu">count</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="population-stratification-assessment-via-pca">Population stratification assessment via PCA<a class="anchor" aria-label="anchor" href="#population-stratification-assessment-via-pca"></a>
</h2>
<p>The following command would compute PCs based on all SNP.</p>
<pre><code><span><span class="va">pcastuff</span> <span class="op">=</span> <span class="va">hl</span><span class="op">$</span><span class="fu">hwe_normalized_pca</span><span class="op">(</span><span class="va">mt17</span><span class="op">$</span><span class="va">GT</span><span class="op">)</span></span></code></pre>
<p>This seems impractical. We have found that with a 64 core machine at
terra.bio, PCA on samples of 1-5% of the 3.8 million loci on T2T chr17
takes 40 minutes. Hail’s code seems good at utilizing all the cores.</p>
<p>We saved the PC scores for PCA based on 38k randomly sampled loci,
and 191k randomly sampled loci. Here’s a simple view of the latter.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pcs_191k</span><span class="op">)</span></span>
<span><span class="fu">graphics</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/graphics/pairs.html" class="external-link">pairs</a></span><span class="op">(</span><span class="va">pcs_191k</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, pch<span class="op">=</span><span class="st">"."</span><span class="op">)</span></span></code></pre></div>
<p><img src="large_t2t_files/figure-html/lk191-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<ol style="list-style-type: decimal">
<li><p>Comment on the gain in information about geographic origin
achieved by using a 5% sample of loci instead of a 1% sample.</p></li>
<li><p>Find the geographic origins of donors of the 1000 genomes
genotypes and bind them to <code>mt17</code> using the methods given in
vignette <code>01_gwas_tut</code>. Use these codes to color the points
in the PCA plot.</p></li>
<li><p>Produce an artificial “phenotype” for these donors via
<code>rnorm(3202,0,1)</code>, bind it to the genotype data, and perform
a naive GWAS. What are the loci most strongly associated with the
artificial phenotype?</p></li>
<li><p>Produce a new artificial phenotype which has some association
with geographic origin of donor, but no association with genotype.
Produce a new naive GWAS, and a third using some of the PCA scores as
covariates. What are the effects of this covariate adjustment on
reasoning about genetic association with the artificial
phenotype?</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="appendix-using-rclone-with-docker-to-get-the-chr17-data">Appendix: using rclone with docker to get the chr17 data<a class="anchor" aria-label="anchor" href="#appendix-using-rclone-with-docker-to-get-the-chr17-data"></a>
</h2>
<p>It can be painful to install and configure rclone. We use a docker
container. Let RC_DATADIR be an environment variable evaluating to an
available folder.</p>
<p>Also, place the text file with contents</p>
<pre><code>[osn]
type = s3
provider = AWS
endpoint = https://mghp.osn.xsede.org
acl = public
no_check_bucket = true</code></pre>
<p>in a file <code>rclone.conf</code> in a folder pointed to by the
environment variable <code>RC_CONFDIR</code>.</p>
<p>Then the following</p>
<pre><code>docker run -v $RC_DATADIR:/data -v $RC_CONFDIR:/config/rclone -t rclone/rclone:latest ls osn:/bir190004-bucket01/Bioc1KGt2t</code></pre>
<p>will list the files with 1KG samples genotyped against the T2T
reference.</p>
<p>Use the rclone <code>copyto</code> command to obtain a local copy of
the zip file <code>t17.zip</code> in the folder pointed to by
<code>$RC_DATADIR</code>:</p>
<pre><code>docker run -v $RC_DATADIR:/data -v $RC_CONFDIR:/config/rclone -t rclone/rclone:latest copyto osn:/bir190004-bucket01/Bioc1KGt2t/t17.zip ./t17.zip</code></pre>
<p>This file should be unzipped in a folder to which the environment
variable <code>HAIL_T2T_CHR17</code> will point.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Vincent Carey.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
